### PR TITLE
allow release creation on commit

### PR DESCRIPTION
**Pull request by earl-warren**
- The code and tests are already there to allow releases to be created on commits.
- This patch modifies the web code to take into account that an commitID could've been passed as target.
- Added unit test.
- Resolves https://codeberg.org/forgejo/forgejo/issues/1196

### CODE REVIEW

File: routers/web/repo/release.go [hunk 1]
```diff
@@ -387,7 +387,9 @@ func NewReleasePost(ctx *context.Context) {
 		return
 	}
 
-	if !ctx.Repo.GitRepo.IsBranchExist(form.Target) {
+	// form.Target can be a branch name or a full commitID.
+	if !ctx.Repo.GitRepo.IsBranchExist(form.Target) &&
+		len(form.Target) == git.SHAFullLength && !ctx.Repo.GitRepo.IsCommitExist(form.Target) {
 		ctx.RenderWithErr(ctx.Tr("form.target_branch_not_exist"), tplReleaseNew, &form)
```
Comment (lunny) [inline routers/web/repo/release.go]: The prompt is not right now. It's not a branch any more.

**Comment by lunny** [comment]
Possible fix #21681

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> Possible fix #21681

Haven't tested, does it support tag now?

**Comment by lunny** [comment]
If `target` is a commit ID, a tag should be created from that commit ID. So we need to create a tag from the title. But now the tag hasn't been created.

图片

**Comment by lunny** [comment]
> the [test](https://github.com/go-gitea/gitea/pull/26366/files#diff-e81fbbfee5202e94b7929e9910feba5b574da0ff5f05bf0b78d5ce6d58a21c1fR251) shows the tag name is used when the target is a commit ID

Yes, tests is PASS. But maybe it's not consistent with the UI operation.

File: tests/integration/release_test.go [hunk 2]
```diff
@@ -31,7 +35,7 @@ func createNewRelease(t *testing.T, session *TestSession, repoURL, tag, title st
 	postData := map[string]string{
 		"_csrf":      htmlDoc.GetCSRF(),
 		"tag_name":   tag,
-		"tag_target": "master",
+		"tag_target": target,
```
Comment (puni9869) [inline tests/integration/release_test.go:38]: 💯 , for removing the hardcoded value.

**Review by puni9869 — APPROVED** [review]

**Pull request closed without merging**